### PR TITLE
Various fixes; Update solc version; Add tests

### PR DESCRIPTION
**Pull request by LukasGasior1**

### CODE REVIEW

File: contracts/ICO/EasyMineIco.sol [hunk 1]
```diff
@@ -70,30 +70,22 @@ contract EasyMineIco {
...
   modifier isPreIco() {
-    if (msg.sender != preIcoAddress) {
-      revert();
-    }
+    require(msg.sender == preIcoAddress);
     _;
   }
 
   modifier isValidPayload() {
-    if (msg.data.length != 0 && msg.data.length != 4 && msg.data.length != 36) {
```
Comment (kosecki123) [inline contracts/ICO/EasyMineIco.sol:94]: Perhaps change to isValidPayload(uint length) {
require(msg.data.length == length)
}

File: contracts/ICO/EasyMineIco.sol [hunk 2]
```diff
@@ -159,35 +150,32 @@ contract EasyMineIco {
...
     atStage(Stages.AuctionStartScheduled)
   {
     uint256 amount = msg.value;
-    if (_receiver == 0 || amount == 0) {
-      revert();
-    }
+    require(_receiver != 0x0);
+    require(amount != 0x0);
+
     if (!wallet.send(amount)) {
```
Comment (kosecki123) [inline contracts/ICO/EasyMineIco.sol]: Change to 

assert(wallet.send(amount)) for more consistence require/assert over revert/throw semantics usage

File: contracts/ICO/EasyMineIco.sol [hunk 3]
```diff
@@ -203,18 +191,16 @@ contract EasyMineIco {
...
     timedTransitions
     atStage(Stages.AuctionStarted)
-    returns (uint amount)
+    returns (uint256 amount)
   {
-    // If a bid is done on behalf of a user via ShapeShift, the receiver address is set.
-    if (receiver == 0) {
-      receiver = msg.sender;
-    }
+    address receiver = msg.sender;
```
Comment (kosecki123) [inline contracts/ICO/EasyMineIco.sol]: No need for extra variable receiver, use msg.sender across the method.

File: contracts/ICO/EasyMineIco.sol [hunk 4]
```diff
@@ -273,23 +257,33 @@ contract EasyMineIco {
...
 
-    if (bids[receiver] == 0) {
-      revert();
-    }
+    require(bids[receiver] != 0);
 
     uint256 tokenCount = bids[receiver] * 10**18 / finalPrice;
     bids[receiver] = 0;
-    easyMineToken.transfer(receiver, tokenCount);
+    if (!easyMineToken.transfer(receiver, tokenCount)) {
```
Comment (kosecki123) [inline contracts/ICO/EasyMineIco.sol]: Use assert...

File: contracts/ICO/EasyMinePreIco.sol [hunk 5]
```diff
@@ -129,7 +116,9 @@ contract EasyMinePreIco {
     // assign the tokens
     // and bid the ICO contract
     totalTokensSold += tokensToReceive;
-    easyMineToken.transfer(msg.sender, tokensToReceive);
+    if (!easyMineToken.transfer(msg.sender, tokensToReceive)) {
```
Comment (kosecki123) [inline contracts/ICO/EasyMinePreIco.sol]: Use assert

File: contracts/ICO/EasyMinePreIco.sol [hunk 6]
```diff
@@ -146,4 +135,15 @@ contract EasyMinePreIco {
     stage = Stages.Closed;
   }
 
+  /* Transfer any ether accidentally left in this contract */
+  function cleanup()
+    public
+    isOwner
+    atStage(Stages.Closed)
+  {
+    if (!owner.send(this.balance)) {
```
Comment (kosecki123) [inline contracts/ICO/EasyMinePreIco.sol]: Use assert

File: contracts/ICO/EasyMineIco.sol (hunk 1 above)
Comment (LukasGasior1) [reply contracts/ICO/EasyMineIco.sol:94]: it varies: 0 when using fallback function, 4 when calling function directly (36 not needed anymore - removed)

**Comment by LukasGasior1** [comment]
@kosecki123 done

**Review by kosecki123 — APPROVED** [review]
Looks good

**Pull request merged**